src/main/java/org/venice/beachfront/bfapi/config/OAuth2Config.java

```diff
@@ -0,0 +1,199 @@
+            new ResourceOwnerPasswordAccessTokenProvider(),
+            new ClientCredentialsAccessTokenProvider()));
+    }
+
+    @RequestMapping("/oauth/start")
+    public String authCode(final Model model, final HttpServletRequest request, final HttpServletResponse response) throws Exception {
+
+    	if (ssoServiceUrl.equals("placeholder")) {
+            model.addAttribute("header", "Warning: You need to bind to the SSO service.");
+            model.addAttribute("warning", "Please bind your app to restore regular functionality");
+            return "configure_warning";
+        }
+
+        // Get information for the User from the /userinfo endpoint available on the SSO Tile
+        final Map<?,?> userInfoResponse = oauth2RestTemplate().getForObject("{ssoServiceUrl}/userinfo", Map.class, ssoServiceUrl);
+        model.addAttribute("ssoServiceUrl",ssoServiceUrl);
+        model.addAttribute("response", toPrettyJsonString(userInfoResponse));
+
+        // Get Access Token
+        final OAuth2AccessToken accessToken = oauth2RestTemplate().getOAuth2ClientContext().getAccessToken();
+        if (accessToken != null) {
+            model.addAttribute("access_token", toPrettyJsonString(parseToken(accessToken.getValue())));
+            model.addAttribute("id_token", toPrettyJsonString(parseToken((String) accessToken.getAdditionalInformation().get("id_token"))));
+        }
+
+        // Confirm User Profile exists; if not, create
+        final String userId = (String) userInfoResponse.get("user_id");
+        final String userName = (String) userInfoResponse.get("user_name");
+        final UserProfile userProfile = getOrCreateUser(userId, userName);
+
+        // Build Response
+        final String uiRedirectUri = UriComponentsBuilder.newInstance().scheme("https").host("beachfront." + this.domain).queryParam("logged_in", "true").build().toUri().toString();
+        final Cookie cookie = createCookie(userProfile.getApiKey(), cookieExpirySeconds);
+
+        response.addCookie(cookie);
+        response.setStatus(HttpStatus.FOUND.value());
+        response.setHeader("Location", uiRedirectUri);
+
+        return "Authentication successful. Redirecting back to application...";
+    }
+
+    @RequestMapping(value="/logout", method = RequestMethod.GET)
+    public String logout(final HttpServletRequest request, final HttpServletResponse response, final Authentication authentication) throws IOException, UserException {
+
+    	final Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        if (auth != null){
+            new SecurityContextLogoutHandler().logout(request, response, auth);
+        }
+		final URL url = new URL(request.getRequestURL().toString());
+		final String urlStr = url.getProtocol() + "://" + url.getAuthority();
+
+		// Server-side invalidation of API Key
+		userProfileService.invalidateKey(userProfileService.getProfileFromAuthentication(authentication));
```

Comment:
Maybe break this into two separate calls?  If an error happens in there, it might be difficult to track down.

Reply:
Done.